rust-alternate/src/connection.rs

```diff
@@ -0,0 +1,256 @@
+use std::{
+    net::SocketAddr,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use anyhow::{Context as _, Result};
+use futures_util::{
+    stream::{SplitSink, SplitStream},
+    Sink, SinkExt, Stream, StreamExt,
+};
+use tokio::net::TcpStream;
+use tokio_tungstenite::{tungstenite::Message as WebsocketMessage, WebSocketStream};
+
+use crate::{
+    game_result::GameStats,
+    message::{Message, MessageType},
+};
+
+pin_project_lite::pin_project! {
+    pub struct ConnectionReadHalf {
+        addr: SocketAddr,
+        #[pin]
+        stream: SplitStream<WebSocketStream<TcpStream>>,
+    }
+}
+
+pin_project_lite::pin_project! {
+    pub struct ConnectionWriteHalf {
+        addr: SocketAddr,
+        #[pin]
+        sink: SplitSink<WebSocketStream<TcpStream>, WebsocketMessage>,
+    }
+}
+
+pin_project_lite::pin_project! {
+    pub struct Connection {
+        addr: SocketAddr,
+        #[pin]
+        read: ConnectionReadHalf,
+        #[pin]
+        write: ConnectionWriteHalf,
+    }
+}
+
+impl Connection {
+    /// Creates new connection read and write halves
+    pub fn new(addr: SocketAddr, stream: WebSocketStream<TcpStream>) -> Self {
+        let (sink, stream) = stream.split();
+
+        Self {
+            addr,
+            read: ConnectionReadHalf { addr, stream },
+            write: ConnectionWriteHalf { addr, sink },
+        }
+    }
+
+    /// Returns the socket address of the connection
+    pub fn get_addr(&self) -> SocketAddr {
+        self.addr
+    }
+
+    /// Waits for `ReadyUp` message from the connection
+    pub async fn wait_ready_up(&mut self) -> Result<()> {
+        self.read.wait_ready_up().await
+    }
+
+    /// Sends `ReadyUp` message to the connection
+    pub async fn ready_up(&mut self) -> Result<()> {
+        self.write.ready_up().await
+    }
+
+    /// Sends `Play` message to the connection
+    pub async fn play(&mut self) -> Result<()> {
+        self.write.play().await
+    }
+
+    /// Splits the connection into read and write halves
+    pub fn split(self) -> (ConnectionReadHalf, ConnectionWriteHalf) {
+        (self.read, self.write)
+    }
+}
+
+impl Stream for Connection {
+    type Item = Result<Message>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.project();
+        let stream = this.read;
+
+        stream.poll_next(cx)
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.read.size_hint()
+    }
+}
+
+impl Sink<Message> for Connection {
+    type Error = anyhow::Error;
+
+    fn poll_ready(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        let this = self.project();
```

Comment:
i assume this is related to `pin_project_lite`.  It seems interesting, and i assume it must help with async and ownership (afaik pinning is important in relation to async only).

Reply:
Yes, this is related to `pin-project`. I had to use it because I decided to implement my own async streams here. Of course there are other ways to do this but I thought this was most straight forward.